presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -723,13 +798,16 @@ public void addIntermediate(GroupByIdBlock groupIdsBlock, Block block)
             // expand array block back into page
             ColumnarArray columnarArray = toColumnarArray(block); // flattens the squashed arrays; so there is no need to flatten block again.
             ColumnarRow columnarRow = toColumnarRow(columnarArray.getElementsBlock()); // contains the flattened array
-            int newPositionCount = columnarRow.getPositionCount(); // number of positions in expanded array (since columnarRow is already flattened)
+            int newPositionCount = columnarRow.getNotNullPositionCount(); // number of positions in expanded array (since columnarRow is already flattened)
             long[] newGroupIds = new long[newPositionCount];
             boolean[] nulls = new boolean[newPositionCount];
             int currentRowBlockIndex = 0;
             for (int groupIdPosition = 0; groupIdPosition < groupIdsBlock.getPositionCount(); groupIdPosition++) {
                 for (int unused = 0; unused < arrayBlock.getBlock(groupIdPosition).getPositionCount(); unused++) {
                     // unused because we are expanding all the squashed values for the same group id
+                    if (arrayBlock.getBlock(groupIdPosition).isNull(unused)) {
+                        break;
+                    }
```

Comment:
Not quite get this newly changed logic. Could we briefly explain?

Reply:
While pre-processing, we first perform a filter on maskChannel. This filter can cause some groupIds to get filtered out completely i.e.. we might end up no rows for the given groupId. In evaluateIntermediate(), we get a groupId as an input and we basically store all the rows for the given groupId in its corresponding `output` block. Since, all rows for the given groupId has been filtered, we basically store a null in the `output` BlockBuilder. This is being done here:

```
if (blockBuilders.get(groupId) == null) {
       singleArrayBlockWriter.appendNull();
}
```

Since, we are storing null for certain groupIds, inside addIntermediate, we can expect output block for a groupId to be null. This logic checks for that case and skip the groupIds for which there are no rows.